### PR TITLE
Support _revinclude

### PR DESCRIPTION
**Pull request by eladiw**
## Description
1. Add support for _revinclude while limiting the number of included items
2. Added revinclude UnitTests
3. Moved some resources creations to the fixture
4. Added OperationOutcome resource if items were truncated.

## Related issues
Addresses [issue #306 ].

## Testing
E2E tests + manual

To verify I ran:

https://localhost:44348/Patient?_revinclude=Observation:subject

And verified the returned sql statement is as expected and returned result set is correct

### CODE REVIEW

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs [hunk 1]
```diff
@@ -344,8 +346,29 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
...
+
+                    if (includeExpression.Reversed)
+                    {
+                        // In case its revinclude, we limit the number of returned items as the resultset size is potentially
+                        // unbounded.
+                        StringBuilder.Append(" TOP (").Append(Parameters.AddParameter(MaxIncludedItems)).Append(") ");
+                    }
+                    else
+                    {
+                        StringBuilder.Append("0 AS IsPartial, "); // can't be partial as we are requesting all includes
```
Comment (tamirkamara) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:358]: Is it possible to keep the same order of the SELECT statement, meaning Sid, IsMatch, IsPartial?

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs [hunk 2]
```diff
@@ -183,7 +183,14 @@ public SearchOptions Create(string compartmentType, string compartmentId, string
             if (searchParams.Include?.Count > 0)
             {
                 searchExpressions.AddRange(searchParams.Include.Select(
-                    q => _expressionParser.ParseInclude(parsedResourceType.ToString(), q))
+                    q => _expressionParser.ParseInclude(parsedResourceType.ToString(), q, false /* reversed */))
```
[the remark this answers is not in the document]
Comment (eladiw) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs:186]: correct. wrong comment. changing to "not reversed"

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs (hunk 1 above)
Comment (eladiw) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs]: done

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Search/BundleFactory.cs [hunk 3]
```diff
@@ -100,6 +100,24 @@ private ResourceElement CreateBundle(SearchResult result, Bundle.BundleType type
...
+
+                // if the query resulted in any item indicating it is partial, add appropriate outcome
+                // as an entry
+                var resource = new OperationOutcome();
+                resource.Issue = new List<OperationOutcome.IssueComponent>();
+                resource.Issue.Add(new OperationOutcome.IssueComponent
+                {
+                    Severity = OperationOutcome.IssueSeverity.Warning,
+                    Code = OperationOutcome.IssueType.Incomplete,
+                });
```
[the remark this answers is not in the document]
Comment (eladiw) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Search/BundleFactory.cs:112]: yes. done

**Comment by brendankowitz** [comment]
I deleted the failing AKS sql instance, looks like going through now

**Comment by johnstairs** [comment]
I have concerns with the performance implications of of using the windowing function in this way, where the count is performed on the entire CTE, ignoring the top. We only need to know if there is one row more than is included, not the total count. 
As an example, on a relatively small DB with 1k synthea patients, I get the following reads:

```
Table 'Resource'. Scan count 2, logical reads 369, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
Table 'Worktable'. Scan count 3, logical reads 8334, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
Table 'ReferenceSearchParam'. Scan count 10, logical reads 36, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
```
If I take out the window function:
```
Table 'Resource'. Scan count 2, logical reads 369, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
Table 'ReferenceSearchParam'. Scan count 10, logical reads 36, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
```

**Comment by eladiw** [comment]
@johnstairs , updated with revised sql query.

I've added a new cte that counts only after the top (max items + 1).

here is an example :

cte2 AS
(
    SELECT DISTINCT TOP (@p2) refSource.ResourceSurrogateId AS Sid1, 0 AS IsMatch 
    FROM dbo.ReferenceSearchParam refSource
    INNER JOIN dbo.Resource refTarget
    ON refSource.ReferenceResourceTypeId = refTarget.ResourceTypeId
        AND refSource.ReferenceResourceId = refTarget.ResourceId
    WHERE refSource.SearchParamId = @p3
        AND refTarget.IsHistory = 0
        AND refSource.IsHistory = 0
        AND refTarget.ResourceTypeId = @p0
        AND refTarget.ResourceSurrogateId IN (SELECT TOP(@p4) Sid1 FROM cte1)
),
cte3 AS
(
    SELECT DISTINCT TOP (@p5) Sid1, IsMatch, CASE WHEN count(*) over() > @p6 then 1 else 0 end AS IsPartial 
    FROM cte2
),
cte4 AS
(
    SELECT Sid1, IsMatch, IsPartial 
    FROM cte1
    UNION ALL
    SELECT Sid1, IsMatch, IsPartial 
    FROM cte3
)

notice that the cte3 is the new layer and that cte2 is not mentioned in cte4.

I also made sure it improved performance.
new version:
Table 'Resource'. Scan count 2, logical reads 375, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
Table 'Worktable'. Scan count 3, logical reads **207**, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.

prev version:
Table 'Resource'. Scan count 2, logical reads 375, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
Table 'Worktable'. Scan count 3, logical reads **3783**, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Search/BundleFactory.cs [hunk 4]
```diff
@@ -100,6 +100,24 @@ private ResourceElement CreateBundle(SearchResult result, Bundle.BundleType type
...
+                if (result.Results.Any(x => x.IsPartial))
+                {
+                    // if the query resulted in any item indicating it is partial, add appropriate outcome
+                    // as an entry
+                    var resource = new OperationOutcome();
+                    resource.Issue = new List<OperationOutcome.IssueComponent>();
+                    resource.Issue.Add(new OperationOutcome.IssueComponent
+                    {
+                        Severity = OperationOutcome.IssueSeverity.Warning,
+                        Code = OperationOutcome.IssueType.Incomplete,
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Search/BundleFactory.cs:113]: I think we should provide details with text to explain what we mean by "incomplete". (`details.text` at least).

**Comment by johnstairs** [comment]
Throwing in some thoughts here: did you consider changing the following from:

``` sql
cte3 AS
(
    SELECT TOP (@p5) Sid1, IsMatch, CASE WHEN count(*) over() > @p6 then 1 else 0 end AS IsPartial
    FROM cte2
),
```

to:

``` sql
cte3 AS
(
    SELECT TOP (@p5) Sid1, IsMatch, CASE WHEN (select count(*) from cte2) > @p6 then 1 else 0 end AS IsPartial
    FROM cte2
),
```

You go from:
```
Table 'Resource'. Scan count 2, logical reads 369, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
Table 'Worktable'. Scan count 3, logical reads 207, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
Table 'ReferenceSearchParam'. Scan count 10, logical reads 36, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
```

to 
```
Table 'Resource'. Scan count 3, logical reads 402, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
Table 'ReferenceSearchParam'. Scan count 20, logical reads 72, physical reads 0, page server reads 0, read-ahead reads 0, page server read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob page server reads 0, lob read-ahead reads 0, lob page server read-ahead reads 0.
```

Fewer total reads, though more from the non-work tables. I think that *when* we store intermediate results in a table variable instead of having CTEs re-evaluated, this approach will be more beneficial. Something like:
``` sql
DECLARE @p0 SmallInt = 103;
DECLARE @p1 Int = 11;
DECLARE @p2 Int = 101;
DECLARE @p3 SmallInt = 895;
DECLARE @p4 Int = 10;
DECLARE @p5 Int = 100;
DECLARE @p6 Int = 100;

SET STATISTICS IO ON;
SET STATISTICS TIME ON;

DECLARE @matches TABLE
(
    Sid1 BIGINT NOT NULL
)

DECLARE @includes TABLE
(
    Sid1 BIGINT NOT NULL
)

;WITH cte0 AS
(
    SELECT ResourceSurrogateId AS Sid1
    FROM dbo.Resource
    WHERE IsHistory = 0
        AND IsDeleted = 0
        AND ResourceTypeId = @p0

)
INSERT INTO @matches
SELECT DISTINCT TOP (@p1) Sid1
    FROM cte0
    OPTION(RECOMPILE)

;WITH cte2 AS
(
    SELECT DISTINCT TOP (@p2) refSource.ResourceSurrogateId AS Sid1, 0 AS IsMatch
    FROM dbo.ReferenceSearchParam refSource
    INNER JOIN dbo.Resource refTarget
    ON refSource.ReferenceResourceTypeId = refTarget.ResourceTypeId
        AND refSource.ReferenceResourceId = refTarget.ResourceId
    WHERE refSource.SearchParamId = @p3
        AND refTarget.IsHistory = 0
        AND refSource.IsHistory = 0
        AND refTarget.ResourceTypeId = @p0
        AND refTarget.ResourceSurrogateId IN (SELECT TOP(@p4) Sid1 FROM @matches  ORDER BY Sid1)
)
INSERT INTO @includes
SELECT TOP (@p5) Sid1
FROM cte2
OPTION(RECOMPILE)


;WITH  cte4 AS
(
    SELECT Sid1, 1 AS IsMatch, 0 AS IsPartial
    FROM @matches
    UNION ALL
    SELECT Sid1, 0 AS IsMatch, CASE WHEN (select count(*) from @includes) > @p6 then 1 else 0 end AS IsPartial
    FROM @includes
)
SELECT r.ResourceTypeId, r.ResourceId, r.Version, r.IsDeleted, r.ResourceSurrogateId, r.RequestMethod, CAST(IsMatch AS bit) AS IsMatch, CAST(IsPartial AS bit) AS IsPartial, r.RawResource
FROM dbo.Resource r
INNER JOIN cte4
ON r.ResourceSurrogateId = cte4.Sid1
ORDER BY r.ResourceSurrogateId ASC
OPTION(RECOMPILE)
```

But you can leave it as-is for now, since we're not doing the table variable refactoring yet. (And *that* approach probably won't be the right one for bulk export :))

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs [hunk 5]
```diff
@@ -390,20 +406,57 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
...
+                        StringBuilder.Append("TOP (").Append(Parameters.AddParameter(MaxIncludedItems)).Append(") ");
+                    }
+
+                    StringBuilder.Append("Sid1, IsMatch, ");
+
+                    if (isRev)
+                    {
+                        StringBuilder.Append("CASE WHEN count(*) over() > ")
+                        .Append(Parameters.AddParameter(MaxIncludedItems))
+                        .AppendLine(" then 1 else 0 end AS IsPartial ");
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:434]: Please capitalize the SQL keywords `THEN`, `ElSE`, `END`

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs [hunk 6]
```diff
@@ -239,7 +241,8 @@ private async Task<SearchResult> SearchImpl(SearchOptions searchOptions, bool hi
                                 null,
                                 null,
                                 null),
-                            isMatch ? SearchEntryMode.Match : SearchEntryMode.Include));
+                            isMatch ? SearchEntryMode.Match : SearchEntryMode.Include,
+                            isPartial));
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs:245]: I think that isPartial belongs more on the `SearchResult` than the individual entries.

**Comment by johnstairs** [comment]
This is looking good! Just a couple of minor comments.

**Comment by eladiw** [comment]
@johnstairs , fixed.

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Search/BundleFactory.cs [hunk 7]
```diff
@@ -101,16 +101,17 @@ private ResourceElement CreateBundle(SearchResult result, Bundle.BundleType type
...
-                    // if the query resulted in any item indicating it is partial, add appropriate outcome
+                    // if the query resulted in a partia; indication, add appropriate outcome
                     // as an entry
                     var resource = new OperationOutcome();
                     resource.Issue = new List<OperationOutcome.IssueComponent>();
                     resource.Issue.Add(new OperationOutcome.IssueComponent
                     {
                         Severity = OperationOutcome.IssueSeverity.Warning,
                         Code = OperationOutcome.IssueType.Incomplete,
+                        Diagnostics = "Number of included items exceeded the maximum count and the returned result was truncated.",
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Search/BundleFactory.cs:114]: Let's add this to a resx.

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Search/BundleFactory.cs [hunk 8]
```diff
@@ -101,16 +101,17 @@ private ResourceElement CreateBundle(SearchResult result, Bundle.BundleType type
                         true);
                 }
 
-                if (result.Results.Any(x => x.IsPartial))
+                if (result.Partial)
                 {
-                    // if the query resulted in any item indicating it is partial, add appropriate outcome
+                    // if the query resulted in a partia; indication, add appropriate outcome
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Search/BundleFactory.cs:106]: typo

File: src/Microsoft.Health.Fhir.Core/Features/Search/SearchResult.cs [hunk 9]
```diff
@@ -68,5 +70,10 @@ public SearchResult(int totalCount, IReadOnlyList<Tuple<string, string>> unsuppo
         /// Gets the continuation token.
         /// </summary>
         public string ContinuationToken { get; }
+
+        /// <summary>
+        /// Gets if this result is partia;.
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.Core/Features/Search/SearchResult.cs:75]: typo. Also we should probably make it clear that this is not related to the continuation token
